File: unpub/lib/src/file_store.dart
import 'dart:io';
import 'package:path/path.dart' as path;
import 'package_store.dart';

class FileStore extends PackageStore {
  String baseDir;

  FileStore(this.baseDir);

  File _getTarballFile(String package, String version) {
    return File(path.join(baseDir, '$package-$version.tar.gz'));
  }

  @override
  Future<void> upload(String package, String version, List<int> content) async {
    var file = _getTarballFile(package, version);
    await file.create(recursive: true);
    await file.writeAsBytes(content);
  }

  @override
  Stream<List<int>> download(String package, String version) {
    return _getTarballFile(package, version).openRead();
  }
}

File: unpub/file_store.go
package unpub

import (
	"io/ioutil"
	"os"
	"path"
)

// FileStore FileStore
type FileStore struct {
	BaseDir string
}

func (f FileStore) getFilePath(name string, version string) string {
	return path.Join(f.BaseDir, name, "-", version, ".tar.gz")
}

// Init Init
func (f FileStore) Init(baseDir string) {
	f.BaseDir = baseDir
	_ = os.Mkdir(f.BaseDir, os.ModeDir)
}

// Upload Upload
func (f FileStore) Upload(name string, version string, data []byte) error {
	p := f.getFilePath(name, version)
	return ioutil.WriteFile(p, data, 0644)
}

// Download Download
func (f FileStore) Download(name string, version string) ([]byte, error) {
	p := f.getFilePath(name, version)
	return ioutil.ReadFile(p)
}
